docs/reference/scripting/using.asciidoc

```diff
@@ -1,198 +1,243 @@
+
+`params`::
+
+    Specifies any named parameters that are passed into the script as
+    variables. <<prefer-params,Use parameters>> instead of hard-coded values to decrease compile time.
+
+[discrete]
+[[hello-world-script]]
+=== Write your first script
+<<modules-scripting-painless,Painless>> is the default scripting language
+for {es}. It is secure, performant, and provides a natural syntax for anyone
+with a little coding experience.
+
+A Painless script is structured as one or more statements and optionally
+has one or more user-defined functions at the beginning. A script must always
+have at least one statement.
+
+The {painless}/painless-execute-api.html[Painless execute API] provides the ability to
+test a script with simple user-defined parameters and receive a result. Let's
+start with a complete script and review its constituent parts.
+
+First, index a document with a single field so that we have some data to work
+with:
 
 [source,console]
--------------------------------------
+----
 PUT my-index-000001/_doc/1
 {
   "my_field": 5
 }
+----
+
+We can then construct a script that operates on that field and run evaluate the
+script as part of a query. The following query uses the
+<<script-fields,`script_fields`>> parameter of the search API to retrieve a
+script valuation. There's a lot happening here, but we'll break it down the
+components to understand them individually. For now, you only need to
+understand that this script takes `my_field` and operates on it.
 
+[source,console]
+----
 GET my-index-000001/_search
 {
   "script_fields": {
     "my_doubled_field": {
-      "script": {
-        "lang":   "expression",
-        "source": "doc['my_field'] * multiplier",
+      "script": { <1>
+        "source": "doc['my_field'].value * params['multiplier']", <2>
         "params": {
           "multiplier": 2
         }
       }
     }
   }
 }
--------------------------------------
-
-[discrete]
-=== Script parameters
-
-`lang`::
-
-    Specifies the language the script is written in.  Defaults to `painless`.
-
-
-`source`, `id`::
+----
+// TEST[continued]
+<1> `script` object
+<2> `script` source
 
-    Specifies the source of the script.  An `inline` script is specified
-    `source` as in the example above. A `stored` script is specified `id`
-    and is retrieved from the cluster state (see <<modules-scripting-stored-scripts,Stored Scripts>>).
+The `script` is a standard JSON object that defines scripts under most APIs
+in {es}. This object requires `source` to define the script itself. The
+script doesn't specify a language, so it defaults to Painless.
 
+[discrete]
+[[prefer-params]]
+=== Use parameters in your script
 
-`params`::
+The first time {es} sees a new script, it compiles it and stores the
+compiled version in a cache. Compilation can be a heavy process. If you're
+passing variables into a script, pass them in as named `params` instead of
```

Comment:
> If you're passing variables into a script, pass them in as named `params` 

To

> Rather than hard coding values in your script, you may pass them as named `params` instead.

Reply:
I'll change this to read:
>Rather than hard-coding values in your script, pass them as named `params` instead.